### ISSUE TITLE
Drag and drop uploading of files through the IDE

### ISSUE THREAD

**Issue by MichaelMauderer** [enso-org/ide#1565]
### Summary
We need a way for people to directly add data files through the UI. This should be done through drag and drop: a file is dragged into the UI window and dropped somewhere in the project. 

### Value
Ease of use with data. 

#### UX

* the dropped file should be placed in the data folder of the project. Name clashes should be resolved automatically. For later: there should be a way to replace an existing file.
* at the point where the data was placed there should be a node that represents the file and allows one to work with it, e.g., `File.new "Path/To/File/Upload"`
* there should be some indication of the file upload going on. Either as part of the new node that is created or as status in the status bar.

#### UI
* we need an api for handling drag and drop event with file data
* we need to be able to receive those events and read the data, and pass it to the controller.

#### Controller

*the graph editor will need an output frp endpoint that allows it to indicate that a file was dropped in a specific location. The controller part should handle the creation of new nodes, status of upload, and engine integration to upload the file.

#### Engine
* there needs to be an API for placing files in the project (data) folder. This should allow to optionally replace an existing file or return an error status if the file already exists.
 

### Acceptance Criteria & Test Cases
Manual test scenario that shows working with files that get dragged and dropped into a repository.

**Comment by iamrecursion** [comment]
The engine component of this is filed at enso-org/enso#1739.

**Comment by farmaazon** [comment]
I have a design question: What if user delete the node created by dropping file, and the file is still uploading? Shall we break upload and delete file?

Perhaps the natural answer is yes, but that leads us to some odd "race condition": if user removes node just before reaching 100%, the uploaded part will be removed; but if they do that a little bit later, the file will remain uploaded.

Of course we could remove the file also when it was entirely uploaded, but that leads us to another question: what if the file was used elsewhere?

**Comment by wdanilo** [comment]
My feelings are as follow (I'd love to know what @MichaelMauderer thinks too):
1. If you start uploading and you delete the node, you cancel the upload as well.
2. In case of a corner case, like deleting node near 100% of upload - whatever. The file can stay on the disk. I'd definitely prefer it to stay rather than delete it.
3. If we will have multiple nodes referring to a file that is being uploaded, all of them should have spinners and the file should cancel uploading when all of such nodes are removed. If this increases complexity a lot, we can implement that in a separate step.

**Comment by MichaelMauderer** [comment]
> If you start uploading and you delete the node, you cancel the upload as well.

Agree. Another alternative would be that there needs another UI for stopping the upload, and you cannot delete the node until the upload has finished or was manually stopped.

> In case of a corner case, like deleting node near 100% of upload - whatever. The file can stay on the disk. I'd definitely prefer it to stay rather than delete it.

We just need to be a bit careful here that we have some way to list all uploaded files in some way. Otherwise, we will end up with lots of orphans, and we want to avoid storing data that the user cannot delete. Especially if we end up having some cloud storage quotes or such things.

> If we will have multiple nodes referring to a file that is being uploaded, all of them should have spinners and the file should cancel uploading when all of such nodes are removed. If this increases complexity a lot, we can implement that in a separate step.

I agree this would make the most sense. Essentially, before we try to load a file, we would need to check if a path is currently being uploaded to and share ownership of the upload between all nodes referencing that path. We only cancel the upload if the last node is deleted.

**Comment by farmaazon** [comment]
Copied the progress bar design by @wdanilo  from the PR, for reference:

Regarding the progress bar design, we want something along these lines (shape only, no text):

https://user-images.githubusercontent.com/1623053/122346937-cafbe480-cf49-11eb-9ec7-878636dade78.mov

However, we do not want it to be circular - we want it to be a double circle capsule. The sliders will be around self ports - marked RED here:

![image](https://user-images.githubusercontent.com/1623053/122349288-69894500-cf4c-11eb-819e-714af11fdf88.png)


In order to create such a progress bar, the following steps should be done:
1. There should be 4 shapes - 2 circles and 2 rectangles:
    ![image](https://user-images.githubusercontent.com/1623053/122347166-09919f00-cf4a-11eb-84e2-17ed14f389f9.png)
2. All the shapes need to be parametrized in a good way - a percentage beginning and percentage end should be possible to be set. We should think about the percentage this way (sorry, there is an error - there should be 75 instead of 100 - I noticed it too late):
    ![image](https://user-images.githubusercontent.com/1623053/122347445-52e1ee80-cf4a-11eb-9d4b-c57a9f17a2c3.png)
3. Let's draw a shape starting at 25% and finishing at 50%:
    ![image](https://user-images.githubusercontent.com/1623053/122347767-af450e00-cf4a-11eb-9da0-e186b740c338.png)
4. Or another shape starting at 5% and finishing at 75%:
     ![image](https://user-images.githubusercontent.com/1623053/122347984-efa48c00-cf4a-11eb-9921-9cc794263f7e.png)
5. This way we can create the mask. Now, we can also create just filled up shape, make it smaller and subtract them together:
6. ![image](https://user-images.githubusercontent.com/1623053/122348418-6e99c480-cf4b-11eb-90f4-b46e5ee9b3e8.png)
7. So it aligns with the mask:
![image](https://user-images.githubusercontent.com/1623053/122348654-ad2f7f00-cf4b-11eb-97f9-ec1d7c2752a8.png)
8. Now we apply the mask:
    ![image](https://user-images.githubusercontent.com/1623053/122348876-e7991c00-cf4b-11eb-872f-e4635aed7865.png)
9. The last part is to place 2 circles at the ending points:
    ![image](https://user-images.githubusercontent.com/1623053/122349061-2202b900-cf4c-11eb-887d-c02539410c64.png)

And we should be done! :)

For now, lets put the slider just on the left side of the node. We will integrate them with self ports once the new ports will be here.
